network/sendScheduler.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"container/heap"
+)
+
+type sendScheduler struct {
+	peers []*wsPeer
+}
+
+func (sched *sendScheduler) add(wp *wsPeer) bool {
+	if wp.inSendScheduler {
+		return false
+	}
+	heap.Push(sched, wp)
+	wp.inSendScheduler = true
+	return true
+}
+func (sched *sendScheduler) next() *wsPeer {
+	if len(sched.peers) == 0 {
+		return nil
+	}
+	wp := sched.peers[0]
+	heap.Pop(sched)
+	return wp
+}
+
+// Len is part of sort.Interface and heap.Interface
+func (sched *sendScheduler) Len() int {
+	return len(sched.peers)
+}
+
+// Less is part of sort.Interface and heap.Interface
+// "Less" is more forward in the queue, higher weight is that.
+func (sched *sendScheduler) Less(i, j int) bool {
+	vi := sched.peers[i]
+	vj := sched.peers[j]
+	if vi.outgoing {
+		// relays connections we have chosen are higher priority
+		if !vj.outgoing {
+			return true
+		}
+	} else if vj.outgoing {
+		// relays connections we have chosen are higher priority
+		return false
+	}
+	// TODO: rank other relays before other nodes
```

Comment:
shouldn't we also check `wn.config.PriorityPeers` like the `peersHeap` does?

Reply:
This setup can cause starvation issues if the prio is only based on the 3 classes of nodes.